plugins/synaprom/fu-self-test.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupd.h>
+
+#include "fu-plugin-private.h"
+#include "fu-test.h"
+
+#include "fu-synaprom-firmware.h"
+
+static void
+fu_test_synaprom_firmware_func (void)
+{
+	const guint8 *buf;
+	gsize sz = 0;
+	g_autofree gchar *filename = NULL;
+	g_autoptr(GBytes) blob1 = NULL;
+	g_autoptr(GBytes) blob2 = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GPtrArray) firmware = NULL;
+
+	filename = fu_test_get_filename (TESTDATADIR, "prometheus-10.01.2861149_eng.pkg");
+	if (filename == NULL) {
+		g_test_skip ("no firmware found");
+		return;
+	}
+	fw = fu_common_get_contents_bytes (filename, &error);
+	g_assert_no_error (error);
+	g_assert_nonnull (fw);
+	buf = g_bytes_get_data (fw, &sz);
+	g_assert_cmpint (sz, ==, 355172);
+	g_assert_cmpint (buf[0], ==, 0x01);
+	g_assert_cmpint (buf[1], ==, 0x00);
+	firmware = fu_synaprom_firmware_new (fw, &error);
+	g_assert_no_error (error);
+	g_assert_nonnull (firmware);
+
+	/* does not exist */
+	blob1 = fu_synaprom_firmware_get_bytes_by_tag (firmware, 0, NULL);
+	g_assert_null (blob1);
+	blob1 = fu_synaprom_firmware_get_bytes_by_tag (firmware,
+						       FU_SYNAPROM_FIRMWARE_TAG_CFG_HEADER,
+						       NULL);
+	g_assert_null (blob1);
+
+	/* header needs to exist */
+	blob1 = fu_synaprom_firmware_get_bytes_by_tag (firmware,
+						       FU_SYNAPROM_FIRMWARE_TAG_MFW_HEADER,
+						       &error);
+	g_assert_no_error (error);
+	g_assert_nonnull (blob1);
+	buf = g_bytes_get_data (blob1, &sz);
+	g_assert_cmpint (sz, ==, 24);
+	g_assert_cmpint (buf[0], ==, 0x41);
+	g_assert_cmpint (buf[1], ==, 0x00);
+	g_assert_cmpint (buf[2], ==, 0x00);
+	g_assert_cmpint (buf[3], ==, 0x00);
+	g_assert_cmpint (buf[4], ==, 0x00);
+
+	/* payload needs to exist */
+	blob2 = fu_synaprom_firmware_get_bytes_by_tag (firmware,
+						       FU_SYNAPROM_FIRMWARE_TAG_MFW_PAYLOAD,
+						       &error);
+	g_assert_no_error (error);
+	g_assert_nonnull (blob2);
+	buf = g_bytes_get_data (blob2, &sz);
+	g_assert_cmpint (sz, ==, 354880);
+	g_assert_cmpint (buf[0], ==, 0x64);
+	g_assert_cmpint (buf[1], ==, 0xe6);
+	g_assert_cmpint (buf[2], ==, 0x03);
+	g_assert_cmpint (buf[3], ==, 0x00);
+	g_assert_cmpint (buf[4], ==, 0x0b);
+}
```

Comment:
If possible it would be good to have a check that parses the version from the binary here too.

Reply:
I wondered about that; although the version is provided in the firmware we don't actually care -- we only use the metainfo-provided value. Or do you mean just to check we're parsing the firmware correctly?

Reply:
I found that type of information useful with dell-dock as a safe guard that things are working properly.  Eg check the version of the firmware in the metainfo matches what's in the binary itself.

Reply:
How do you get the version in the metainfo file? You just get a GBytes blob and the flags...?

Reply:
I'm not inclined to make this more complicated, I mean, I got burned with the whole version format thing. For instance, if the metainfo file says "1.2.3" and the firmware says "01.02.03" is that the same thing?